File: hash_generator.go
// hash_generator.go — Go версия

package main

import (
	"crypto/md5"
	"encoding/base64"
	"encoding/hex"
	"flag"
	"fmt"
	"io"
	"os"
)

func computeHash(data []byte) string {
	h := md5.New()
	h.Write(data)
	return hex.EncodeToString(h.Sum(nil))
}

func hashFile(filename string) (string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()

	h := md5.New()
	buf := make([]byte, 8192)
	for {
		n, err := file.Read(buf)
		if err != nil && err != io.EOF {
			return "", err
		}
		if n == 0 {
			break
		}
		h.Write(buf[:n])
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

func main() {
	format := flag.String("format", "hex", "Формат вывода (hex, base64)")
	filePath := flag.String("file", "", "Файл для хеширования")
	output := flag.String("output", "", "Сохранить хеш в файл")
	compare := flag.String("compare", "", "Сравнить с эталонным хешем")
	flag.Parse()

	fmt.Println("\x1b[36m🔐 Hash Generator (MD5) (Go)\x1b[0m")
	fmt.Println("Алгоритм: MD5")

	var hexDigest string
	var err error

	if *filePath != "" {
		fmt.Printf("📂 Хеширование файла: %s\n", *filePath)
		hexDigest, err = hashFile(*filePath)
		if err != nil {
			fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
			os.Exit(1)
		}
	} else if flag.NArg() > 0 {
		input := flag.Arg(0)
		fmt.Printf("📝 Входные данные: %s\n", input)
		hexDigest = computeHash([]byte(input))
	} else {
		// Чтение из stdin
		fmt.Println("📝 Чтение из STDIN (Ctrl+D для окончания)")
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			data, err := io.ReadAll(os.Stdin)
			if err != nil {
				fmt.Printf("\x1b[31m❌ Ошибка чтения: %v\x1b[0m\n", err)
				os.Exit(1)
			}
			hexDigest = computeHash(data)
		} else {
			fmt.Println("Введите строку (Enter для завершения):")
			var input string
			fmt.Scanln(&input)
			if input == "" {
				fmt.Println("\x1b[33m⚠️ Пустой ввод.\x1b[0m")
				os.Exit(1)
			}
			hexDigest = computeHash([]byte(input))
		}
	}

	var result string
	if *format == "base64" {
		decoded, _ := hex.DecodeString(hexDigest)
		result = base64.StdEncoding.EncodeToString(decoded)
	} else {
		result = hexDigest
	}

	fmt.Printf("\x1b[32mХеш (%s):\x1b[0m\n", *format)
	fmt.Println(result)

	if *compare != "" {
		if result == *compare {
			fmt.Println("\x1b[32m✅ Хеши совпадают!\x1b[0m")
		} else {
			fmt.Println("\x1b[31m❌ Хеши не совпадают!\x1b[0m")
		}
	}

	if *output != "" {
		err := os.WriteFile(*output, []byte(result+"\n"), 0644)
		if err != nil {
			fmt.Printf("\x1b[31m❌ Ошибка сохранения: %v\x1b[0m\n", err)
		} else {
			fmt.Printf("\x1b[32m💾 Сохранено в %s\x1b[0m\n", *output)
		}
	}
}

File: hash_generator.rb
# hash_generator.rb — Ruby версия

require 'digest'
require 'optparse'
require 'base64'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby hash_generator.rb [options] [input]"
  opts.on("-f", "--format FORMAT", "Формат (hex, base64)") { |f| options[:format] = f }
  opts.on("-F", "--file FILE", "Файл для хеширования") { |f| options[:file] = f }
  opts.on("-o", "--output FILE", "Сохранить хеш в файл") { |o| options[:output] = o }
  opts.on("-c", "--compare HASH", "Сравнить с эталонным хешем") { |c| options[:compare] = c }
end.parse!

format = options[:format] || 'hex'
file_path = options[:file]
output = options[:output]
compare = options[:compare]
input = ARGV[0]

puts "\e[36m🔐 Hash Generator (MD5) (Ruby)\e[0m"
puts "Алгоритм: MD5"

def compute_hash(data)
  Digest::MD5.hexdigest(data)
end

def hash_file(filename)
  Digest::MD5.file(filename).hexdigest
end

hex_digest = if file_path
  puts "📂 Хеширование файла: #{file_path}"
  hash_file(file_path)
elsif input
  puts "📝 Входные данные: #{input}"
  compute_hash(input)
else
  puts "📝 Чтение из STDIN (Ctrl+D для окончания)"
  data = STDIN.read
  if data.empty?
    puts "\e[33m⚠️ Пустой ввод.\e[0m"
    exit 1
  end
  compute_hash(data)
end

result = if format == 'base64'
  Base64.strict_encode64([hex_digest].pack('H*'))
else
  hex_digest
end

puts "\e[32mХеш (#{format}):\e[0m"
puts result

if compare
  if result == compare
    puts "\e[32m✅ Хеши совпадают!\e[0m"
  else
    puts "\e[31m❌ Хеши не совпадают!\e[0m"
  end
end

if output
  File.write(output, result + "\n")
  puts "\e[32m💾 Сохранено в #{output}\e[0m"
end
